### ISSUE TITLE
Database operations within a task within a transaction in test mode hang indefinitely

### ISSUE THREAD

**Issue by dkulchenko** [elixir-ecto/ecto_sql#122]
### Environment

* Elixir version (elixir -v): 1.9.0
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): PostgreSQL 11
* Ecto version (mix deps): 3.1.7
* Database adapter and version (mix deps): postgrex 0.14.3
* Operating system: Ubuntu 18.04 LTS

### Current behavior

The following test case in a blank Phoenix project will cause the inner task to hang indefinitely (until Ecto timeout):

```elixir
 test "should allow creating an item from a task within a transaction", %{conn: _conn} do
    Repo.transaction(fn ->
      Task.async(fn ->
        %Item{}
        |> Repo.insert!()
      end)
      |> Task.await()
    end)
  end
```

The error is:

```
  1) test failing test case (EctoFailCaseWeb.PageControllerTest)
     test/ecto_fail_case_web/controllers/page_controller_test.exs:4
     ** (EXIT from #PID<0.363.0>) an exception was raised:
         ** (DBConnection.ConnectionError) connection not available and request was dropped from queue after 978ms. You can configure how long requests wait in the queue using :queue_target and :queue_interval. See DBConnection.start_link/2 for more information
             (ecto_sql) lib/ecto/adapters/sql.ex:621: Ecto.Adapters.SQL.raise_sql_call_error/1
             (ecto) lib/ecto/repo/schema.ex:649: Ecto.Repo.Schema.apply/4
             (ecto) lib/ecto/repo/schema.ex:262: anonymous fn/15 in Ecto.Repo.Schema.do_insert/4
             (ecto) lib/ecto/repo/schema.ex:164: Ecto.Repo.Schema.insert!/4
             (elixir) lib/task/supervised.ex:90: Task.Supervised.invoke_mfa/2
             (elixir) lib/task/supervised.ex:35: Task.Supervised.reply/5
             (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```

Switching the Ecto sandbox to shared mode and toggling async to `false` doesn't seem to fix the issue. However, the same code in a non-test environment (e.g. IEx running in dev mode) succeeds without issue.

Full example project (essentially just a `mix phx.new` with the above test case and a single schema) at: https://github.com/dkulchenko/ecto-fail-case

**Comment by michalmuskala** [comment]
In general, transactions cannot span across processes - either during tests with sandbox or regular operation. We could potentially improve the error here, though.

**Comment by dkulchenko** [comment]
That makes sense, but in this case I'm not intending for the outer transaction to be reused in the task process - it's meant to be its own stream of work (and if I understand correctly, this is what happens in a non-test environment, the task would check out a separate connection).

But spawning that task from within an open transaction during a test prevents it from executing at all, which doesn't seem right.

**Comment by josevalim** [comment]
@dkulchenko when using the sandbox, the connections are either explicitly assigned or shared. So one of two things will happen in your tasks:

1. You have a single connection, which you are trying to use across all processes, which causes it to fail

2. The task does not have a connection assigned to it, which causes it to fail

We added automatically sharing for tasks in a sandbox in recent versions, which means you are triggering 1. But without 1, you would just trigger 2.

**Comment by josevalim** [comment]
In any case, here is a PR that improves the error message in this case: https://github.com/elixir-ecto/db_connection/pull/200

**Comment by josevalim** [comment]
Closing in favor of the PR above. :)

**Comment by dkulchenko** [comment]
@josevalim Understood - thank you for the explanation :) Will rewrite things a bit so that the one shared connection can be used.

**Comment by igsp7** [comment]
I am facing the exact same situation. @dkulchenko did you manage to make your Tasks use the shared connection?

**Comment by dkulchenko** [comment]
I pulled up the old project that I ran into this with and it looks like my solution was to untangle the cyclic dependency. 

Basically, the test case above hangs because the async task is trying to use the sandbox connection, but the parent task has it checked out for the transaction and can't release it (because the transaction call itself is blocking on the async task completing). 

The real world case I was encountering this in was calling an audit function that needed to (in the background) add some auditing rows to the DB (and this function was itself called from within a parent transaction). I ended up moving the auditing code to run in a Task.Supervisor and not blocking on the call inside the parent transaction. 

I also added some "real" DB tests that run against a test DB with sandbox mode disabled and exercise the code using the normal path.

Looking through the old code, it looks like my approaches were essentially to have a wrapper MyProject.Tasks module that selectively inlined async tasks in test (to not spawn a separate task at all) or to untangle dependencies using non-blocking/non-immediately-awaited tasks, keeping in mind there's only one connection available at a time in the test sandbox. This can be supplemented with non-sandboxed tests as needed to fill in any coverage gaps.

This was really early in my Elixir learning so it's possible that newer releases of Ecto have mitigated this in other ways, or perhaps there's better approaches.
